pkg/health/health.go

```diff
@@ -0,0 +1,117 @@
+	Timeout metav1.Duration `json:"timeout"`
+}
+
+// Resource kubernetes object to have checks run against
+// An object has checks evaluated against the object
+type Resource struct {
+	Group      string   `json:"group"`
+	APIVersion string   `json:"apiversion"`
+	Resource   string   `json:"resource"`
+	Name       string   `json:"name"`
+	Namespace  string   `json:"namespace"`
+	Checks     []*Check `json:"checks"`
+}
+
+// conditionExpression wrap condition function using a closure allowing the event call back to have an expression
+func conditionExpression(expression string) k8s.ConditionFunction {
+	return func(event watch.Event, obj *unstructured.Unstructured) (bool, error) {
+		// compile expression with object, Env help with typing during evaluation
+		pgrm, err := expr.Compile(expression, expr.Env(obj.Object), expr.AsBool())
+		if err != nil {
+			return false, err
+		}
+		// check against the object
+		output, err := expr.Run(pgrm, obj.Object)
+		if err != nil {
+			return false, err
+		}
+		return output.(bool), nil
+	}
+}
+
+// Check run wait on resource until expression passes
+// note at the moment it doesn't track any scycess/fails
+func (h *Resource) Check(clientFactory factory.Factory, namespace string) (bool, error) {
+	clientMgr := k8s.NewK8sClientMgr(clientFactory)
+	namespace, err := clientMgr.Namespace()
+	if err != nil {
+		return false, err
+	}
+	gvr := schema.GroupVersionResource{
+		Group:    h.Group,
+		Version:  h.APIVersion,
+		Resource: h.Resource,
+	}
+	passed := true
+	for _, check := range h.Checks {
+		// TODO WatchEventsForCondition should use a context
+		met, err := clientMgr.WatchEventsForCondition(int(check.Timeout.Seconds()), namespace, h.Name, gvr, conditionExpression(check.Expression))
+		if err != nil {
+			return false, err
+		}
+		if !met {
+			passed = false
+		}
+	}
+	return passed, nil
+}
+
+// V1AlphaHealthSpec HealthSpec
+type V1AlphaHealthSpec struct {
+	Resources []Resource      `json:"resources"`
+	Timeout   metav1.Duration `json:"timeout"`
+}
+
+// Health is bunch of resources that should be checked together as a group
+type Health struct {
+	Spec               V1AlphaHealthSpec `json:"spec"`
+	Name               string            `json:"name"`
+	Metadata           metav1.ObjectMeta `json:"metadata"`
+	healthy, unhealthy []string
+}
+
+// CheckResources wait until all resources checks passed of have been exhausted
+// return true if all resources passed checks
+func (h *Health) CheckResources(client k8s.ClientMgr) (bool, error) {
+	// track reuslts
+	var err error
+	for _, r := range h.Spec.Resources {
+		ns, err := client.Namespace()
+		if err != nil {
+			return false, err
+		}
+		healthy, err := r.Check(client, ns)
+		if !errors.Is(err, k8s.ErrWatchTimeout) && err != nil {
+			err = errors.Wrapf(err, "%s checks failed", r.Name)
+
+		}
+		if !healthy {
+			h.unhealthy = append(h.unhealthy, r.Name)
+			continue
+		}
+		if healthy {
+			h.healthy = append(h.healthy, r.Name)
+		}
```

Comment:
don't need second `if` after `continue`
```suggestion
		if !healthy {
			h.unhealthy = append(h.unhealthy, r.Name)
			continue
		}
		h.healthy = append(h.healthy, r.Name)
		
```